singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/AuditEventKafkaSender.java

```diff
@@ -0,0 +1,426 @@
+
+  /**
+   *  If sending out to one partition fails, this partition is added to set;
+   *  If sending out to one partition succeeds, this partition is removed if it was added before.
+   */
+  private Set<Integer> badPartitions = ConcurrentHashMap.newKeySet();
+
+  /**
+   * For each event (identified by LoggingAuditHeaders, key of the map), track the number of tries
+   * for sending to Kafka. Each try will choose a different and partition that is not in the
+   * badPartitions set.  When event is send out successfully or dropped, the corresponding entry in
+   * this map is removed.
+   */
+
+  private Map<LoggingAuditHeaders, Integer> eventTriedCount = new ConcurrentHashMap<>();
+
+  public AuditEventKafkaSender(KafkaSenderConfig config,
+                               LinkedBlockingDeque<LoggingAuditEvent> queue,
+                               LoggingAuditStage stage, String host, String name) {
+    this.topic = config.getTopic();
+    this.queue = queue;
+    this.stage = stage;
+    this.host = host;
+    this.name = name;
+  }
+
+
+  public KafkaProducer<byte[], byte[]> getKafkaProducer() {
+    return kafkaProducer;
+  }
+
+  public void setKafkaProducer(KafkaProducer<byte[], byte[]> kafkaProducer) {
+    this.kafkaProducer = kafkaProducer;
+  }
+
+  public int getAlternatePartition(int numOfPartitions) {
+    int randomPartition = 0;
+    int trials = 10;
+    while (trials > 0) {
+      trials -= 1;
+      randomPartition = ThreadLocalRandom.current().nextInt(numOfPartitions);
+      if (!badPartitions.contains(randomPartition)) {
+        break;
+      }
+    }
+    return randomPartition;
+  }
+
+  private void refreshPartitionIfNeeded() {
+    // refresh every 5 mins
+    if (System.currentTimeMillis() - lastTimeUpdate > 1000 * 300) {
+      try {
+        partitionInfoList = this.kafkaProducer.partitionsFor(topic);
+        lastTimeUpdate = System.currentTimeMillis();
+        OpenTsdbMetricConverter
+            .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_KAFKA_PARTITIONS_REFRESH_COUNT, 1,
+                "host=" + host, "stage=" + stage.toString());
+      } catch (Exception e) {
+        OpenTsdbMetricConverter
+            .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_SENDER_KAFKA_PARTITIONS_REFRESH_ERROR, 1,
+                "host=" + host, "stage=" + stage.toString());
+      }
+    }
+  }
+
+  /**
+   *  Sender dequeue LoggingAuditEvents and send out to Kafka cluster. If send out to one partition
+   *  fails, it will choose another partition. For each event, it will try at most
+   *  umOfPartitionsToTrySending times (default 3 times) before dropping this events. Metrics are
+   *  used to track the queue size and usuage, number of events sent out to Kafka successfully, and
+   *  the number of events dropped.
+   */
+  @Override
+  public void run() {
+    LoggingAuditEvent event = null;
+    ProducerRecord<byte[], byte[]> record;
+    byte[] value = null;
+    refreshPartitionIfNeeded();
```

Comment:
Note that.  In refreshPartitionIfNeeded(), there is try...catch block and failing to refresh the List<PartitionInfo> will not break/exit the application; metrics are used to track the refresh success / failures.  This is the expected behavior in the while loop (LINE 227) because we can still reply on the old (not refreshed) List<PartitionInfo> to process more events.

In line 221, this call is to initialize List<PartitionInfo> and a Preconditions checking is right after this line to make sure, the list is not empty in order to entry the WHILE loop.  I feel it is better to keep it.

Reply:
what is the value of having `Preconditions.checkArgument(partitionInfoList.size() > 0)` checking?   if the condition fails, will the program throws exception and terminate?   in general, we should only keep the code that is really needed. otherwise it will cause confusion to other readers later.

Reply:
removed line 221 to line 223.  The refresh method in the WHILE loop is enough. Made the change in the latest update of this PR.